levels.go

```diff
@@ -573,36 +573,59 @@ func (s *levelsController) compactBuildTables(
 			// Do not discard entries inserted by merge operator. These entries will be
 			// discarded once they're merged
 			if version <= discardTs && vs.Meta&bitMergeEntry == 0 {
-				// Keep track of the number of versions encountered for this key. Only consider the
-				// versions which are below the minReadTs, otherwise, we might end up discarding the
-				// only valid version for a running transaction.
-				numVersions++
-
-				// Keep the current version and discard all the next versions if
-				// - The `discardEarlierVersions` bit is set OR
-				// - We've already processed `NumVersionsToKeep` number of versions
-				// (including the current item being processed)
-				lastValidVersion := vs.Meta&bitDiscardEarlierVersions > 0 ||
-					numVersions == s.kv.opt.NumVersionsToKeep
-
-				if isDeletedOrExpired(vs.Meta, vs.ExpiresAt) || lastValidVersion {
-					// If this version of the key is deleted or expired, skip all the rest of the
-					// versions. Ensure that we're only removing versions below readTs.
+				// When `NumVersionsToKeep == 1` the skipKey should always be set
+				// and there is no need for `lastValidVersion`. The only required check
+				// is to determine if a deleted key should be written due to `hasOverlap`.
+				if s.kv.opt.NumVersionsToKeep == 1 {
```

Comment:
All the changes in `levels.go` file can be replaced with the following
```diff
diff --git a/levels.go b/levels.go
index 764fcd9..0121650 100644
--- a/levels.go
+++ b/levels.go
@@ -578,6 +578,16 @@ func (s *levelsController) compactBuildTables(
 				// only valid version for a running transaction.
 				numVersions++
 
+				if isDeletedOrExpired(vs.Meta, vs.ExpiresAt) && !hasOverlap &&
+					s.kv.opt.NumVersionsToKeep == 1 {
+					skipKey = y.SafeCopy(skipKey, it.Key())
+
+					// If no overlap, we can skip all the versions, by continuing here.
+					numSkips++
+					updateStats(vs)
+					continue // Skip adding this key.
+				}
+
 				// Keep the current version and discard all the next versions if
 				// - The `discardEarlierVersions` bit is set OR
 				// - We've already processed `NumVersionsToKeep` number of versions
```

I tried this and everything seems to be working fine. I'll push some more tests on this PR to ensure we've covered all the cases. We have very few tests for compaction which were added recently.